driver/esp_camera.c

```diff
@@ -141,28 +149,90 @@ static esp_err_t camera_probe(const camera_config_t *config, camera_model_t *out
      * Read sensor ID
      */
     sensor_id_t *id = &s_state->sensor.id;
+    do {
+        if (slv_addr == OV2640_SCCB_ADDR) {
+            SCCB_Write(slv_addr, 0xFF, 0x01);//bank sensor
+            id->PID = SCCB_Read(slv_addr, REG_PID);
+            id->VER = SCCB_Read(slv_addr, REG_VER);
+            id->MIDL = SCCB_Read(slv_addr, REG_MIDL);
+            id->MIDH = SCCB_Read(slv_addr, REG_MIDH);
+            if (OV2640_PID == id->PID) {
+                break;
+            }
 
-    if (slv_addr == OV2640_SCCB_ADDR || slv_addr == OV7725_SCCB_ADDR) {
-        SCCB_Write(slv_addr, 0xFF, 0x01);//bank sensor
-        id->PID = SCCB_Read(slv_addr, REG_PID);
-        id->VER = SCCB_Read(slv_addr, REG_VER);
-        id->MIDL = SCCB_Read(slv_addr, REG_MIDL);
-        id->MIDH = SCCB_Read(slv_addr, REG_MIDH);
-    } else if (slv_addr == OV5640_SCCB_ADDR || slv_addr == OV3660_SCCB_ADDR) {
-        id->PID = SCCB_Read16(slv_addr, REG16_CHIDH);
-        id->VER = SCCB_Read16(slv_addr, REG16_CHIDL);
-    } else if (slv_addr == NT99141_SCCB_ADDR) {
-        SCCB_Write16(slv_addr, 0x3008, 0x01);//bank sensor
-        id->PID = SCCB_Read16(slv_addr, 0x3000);
-        id->VER = SCCB_Read16(slv_addr, 0x3001);
-        if (config->xclk_freq_hz > 10000000) {
-            ESP_LOGE(TAG, "NT99141: only XCLK under 10MHz is supported, and XCLK is now set to 10M");
-            s_state->sensor.xclk_freq_hz = 10000000;
         }
-    }
-    vTaskDelay(10 / portTICK_PERIOD_MS);
+        if (slv_addr == OV7725_SCCB_ADDR) {
```

Comment:
Given that some sensors have the same I2C addresses, the following code will execute a few times. For example, 7725 and 7670 have the same I2C address and the same registers for checking ID. Code should read from I2C once and checks who's PID it got, instead of reading the same thing a few times and checking one by one. Please organize this portion better :)

Reply:
It is inevitable to read many times, because the address of PID register of different sensors is not the same, we always have to try one by one

Reply:
I do not agree here. I think there is a much better way to handle this. If the case is that you got a GC sensor with address like OV sensor and you get wrong PID from the OV check, you can also try the GC sensors and see if it matches one of them. Same goes for OV56 and OV36. Same address, same registers.

Reply:
My idea is to make each sensor independent. No matter how many sensors there are, each sensor only detects whether it matches its own PID(such as api gc032a_detect()). This makes adding sensors more convenient

Reply:
I do not see how that improves adding a new sensor. What we are discussing is the current sensors. If you have attached GC0308, you will go through 4 transactions before you check the correct sensor. For 7670 you will ask 3 times the same exact thing and check only on the third time if the value is correct. Comments are enough to signify what is going on. Also you should stay consistent with how camera PID is read. Do not use functions :)

Reply:
You're right. There will be a few redundant SCCB transactions, which will only take a little more time. The advantage of this method is that it does not need to consider which sensors have the same SCCB address and which are different, and which sensors have the same PID register address and which are different.

I want to add a detect function to all sensors to check if the PID matches itself. Return what the sensor should do to the sensor :)

Reply:
We already know that they have the same address and same registers. We already support them. This looks like over-abstraction. The only benefit is that the code is slightly more readable, which could be mitigated by adding proper comments. As a matte of fact, we should also skip reading VER, MIDL and MIDH. We do not use them anywhere

Reply:
VER, MIDL and MIDH do not seem to be used because many sensors do not have these parameters

Reply:
yes, not needed :)

Reply:
I'll take some time to improve on this part :frowning:

Reply:
Please add error log when PID is not recognized to print out PID+I2C address. It can help when adding new sensors